internal/worker/sshsession/worker.go

```diff
@@ -0,0 +1,364 @@
+		return errors.Trace(err)
+	}
+
+	return nil
+}
+
+// cleanupPublicKey finds the fingerprint of the provided key and attempts to delete the key
+// from authorized_keys2.
+func (w *sshSessionWorker) cleanupPublicKey(ephemeralPublicKey string) error {
+	fingerprint, _, err := jujussh.KeyFingerprint(ephemeralPublicKey)
+	if err != nil {
+		w.logger.Errorf("failed to get fingerprint of ephemeral public key: %v", err)
+		return errors.Trace(err)
+	}
+
+	if err := jujussh.DeleteKeys(ControllerSSHUser, AuthorizedKeysFile, true, fingerprint); err != nil {
+		w.logger.Errorf("Error deleting ephemeral public key: %v", err)
+		return errors.Trace(err)
+	}
+
+	return nil
+}
+
+type ConnectionGetter interface {
+	GetSSHConnection(password, ctrlAddress string) (net.Conn, error)
+	GetSSHDConnection() (net.Conn, error)
+}
+
+type connectionGetter struct {
+	logger Logger
+}
+
+// NewConnectionGetter returns a struct capable of initating SSH connections to two places.
+//
+//  1. The controller's SSH server
+//  2. The local SSHD on the machine
+//
+// The consumer is expect to pipe these connections together.
+func NewConnectionGetter(l Logger) *connectionGetter {
+	return &connectionGetter{l}
+}
+
+// GetSSHConnection initiates an SSH connection to the target ctrlAddress.
+func (w *connectionGetter) GetSSHConnection(password, ctrlAddress string) (net.Conn, error) {
+	// TODO(ale8k): Watch will return host key in subsequent PR.
+	sshConfig := &ssh.ClientConfig{
+		User:            ControllerSSHUser,
+		Auth:            []ssh.AuthMethod{ssh.Password(password)},
+		HostKeyCallback: ssh.InsecureIgnoreHostKey(), // TODO(ale8k): Fill in host key here
+	}
+
+	client, err := ssh.Dial("tcp", ctrlAddress, sshConfig)
+	if err != nil {
+		w.logger.Errorf("failed to dial controller %q: %v", ctrlAddress, err)
+		return nil, errors.Trace(err)
+	}
+
+	conn, err := client.Dial("tcp", ctrlAddress)
+	if err != nil {
+		w.logger.Errorf("failed to dial controller %q: %v", ctrlAddress, err)
+		return nil, errors.Trace(err)
+	}
+
+	return conn, nil
+}
+
+// GetSSHConnection performs a stand TCP dial to the SSHD running on the machine.
+func (w *connectionGetter) GetSSHDConnection() (net.Conn, error) {
+	etcPath := "/etc/ssh/sshd_config"
+	openSSHPath := "/usr/share/openssh/sshd_config"
+	port := w.getLocalSSHPort(etcPath, openSSHPath)
+	u, err := url.Parse("localhost" + port)
+	if err != nil {
+		w.logger.Errorf("failed to parse local sshd port: %v", err)
+		return nil, errors.Trace(err)
+	}
+
+	localSSHD, err := net.Dial("tcp", u.String())
+	if err != nil {
+		w.logger.Errorf("Failed to connect to local sshd: %v", err)
+		return nil, errors.Trace(err)
+	}
+	return localSSHD, nil
+}
+
+// getLocalSSHPort parses the local sshd_config file to get the port sshd is listening on.
+// It will try all the provided paths.
+// If it cannot get it for any reason, it'll log the error and return the default port 22.
+func (w *connectionGetter) getLocalSSHPort(filePaths ...string) string {
```

Comment:
Left a comment about this in the test file. I don't think we need to do this.

Reply:
As discussed, we'll leave it.